File: JavaRecyclerViewApp/app/src/main/java/com/example/javarecyclerviewapp/RecyclerViewAdapter.java
package com.example.javarecyclerviewapp;

import android.content.Context;
import android.graphics.Color;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {

    private Context context;
    private ArrayList<ListItems> listItems = new ArrayList<>();
    private ArrayList<ListItems> fullList = new ArrayList<>();
    private boolean filtered = false;

    public RecyclerViewAdapter(Context contxt, ArrayList<ListItems> listItms) {
        this.context = contxt;
        this.listItems = listItms;
        this.fullList = listItms;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_items, parent, false);
        ViewHolder viewHolder = new ViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            holder.text.setText(listItems.get(position).getText());
            Picasso.get().load(listItems.get(position).getImage()).into(holder.image);
            holder.layout.setBackgroundColor(listItems.get(position).getColor());
    }


    @Override
    public int getItemCount() {
        return listItems.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder{

        ImageView image;
        TextView text;
        RelativeLayout layout;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            image = itemView.findViewById(R.id.imageView);
            text = itemView.findViewById(R.id.textView);
            layout = itemView.findViewById(R.id.relativeLayout);
            itemView.setOnClickListener(new View.OnClickListener() {

                @Override
                public void onClick(View v) {
                    if (filtered == false) {
                        filter(listItems.get(getAdapterPosition()).getColor());
                    }
                    else {
                        listItems = fullList;
                        filtered = false;
                        notifyDataSetChanged();
                    }
                }
            });
        }

    }
    private void filter(int color)
    {
        ArrayList<ListItems> filteredList = new ArrayList<>();

        for (ListItems item : listItems) {
            if (item.getColor() == color) {
                filteredList.add(item);
            }
        }
        listItems = filteredList;
        filtered = true;
        notifyDataSetChanged();
    }
}

File: JavaRecyclerViewApp/app/src/main/java/com/example/javarecyclerviewapp/RecyclerViewAdapter.kt
package com.example.javarecyclerviewapp

import android.content.Context
import android.graphics.Color
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.squareup.picasso.Picasso

import java.util.ArrayList

class RecyclerViewAdapter(private val context: Context, listItms: ArrayList<ListItems>) : RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder>() {
    private var listItems = ArrayList<ListItems>()
    private val fullList: ArrayList<ListItems>
    private var filtered = false

    init {
        this.listItems = listItms
        this.fullList = listItms
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        val view = LayoutInflater.from(parent.context).inflate(R.layout.list_items, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.text.text = listItems[position].text
        Picasso.get().load(listItems[position].image).into(holder.image)
        holder.layout.setBackgroundColor(listItems[position].color)
    }


    override fun getItemCount(): Int {
        return listItems.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        internal var image: ImageView = itemView.findViewById(R.id.imageView)
        internal var text: TextView = itemView.findViewById(R.id.textView)
        internal var layout: RelativeLayout = itemView.findViewById(R.id.relativeLayout)

        init {
            itemView.setOnClickListener {
                if (!filtered) {
                    filter(listItems[adapterPosition].color)
                } else {
                    listItems = fullList
                    filtered = false
                    notifyDataSetChanged()
                }
            }
        }

    }

    private fun filter(color: Int) {
        val filteredList = ArrayList<ListItems>()

        for (item in listItems) {
            if (item.color == color) {
                filteredList.add(item)
            }
        }
        listItems = filteredList
        filtered = true
        notifyDataSetChanged()
    }
}
